### PR TITLE
Add URI based allowPrefexList and denyPrefixList

### PR DESCRIPTION
**Pull request by nishantmonu51**
Description
Right now there is no support to restrict druid based ingestion to deny ingestion from specific paths.

This PR adds a new config `druid.ingestion.uri.allowPrefexList` and `druid.ingestion.uri.denyPrefexList` which allows/deny configured URI prefixes from being accessed through druid ingestion.

Default behavior is to allow ingestion from all paths. 

This PR has:

[*] been self-reviewed.
[*] added documentation for new or modified features or behaviors.
[*] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for code coverage is met.

### CODE REVIEW

File: core/src/main/java/org/apache/druid/data/input/impl/LocalInputSource.java
Comment (a2l007) [inline core/src/main/java/org/apache/druid/data/input/impl/LocalInputSource.java:82]: Maybe out of this PR's scope but it could be useful to support pattern based validation for `filter`.

**Review by a2l007 — APPROVED** [review]
+1 after CI and typo fixes

File: core/src/main/java/org/apache/druid/data/input/impl/InputSourceSecurityConfig.java
Comment (himanshug) [reply core/src/main/java/org/apache/druid/data/input/impl/InputSourceSecurityConfig.java:45]: +1 , maybe introduce a serde test that would catch this.

**Comment by himanshug** [comment]
LGTM overall

File: core/src/main/java/org/apache/druid/data/input/impl/InputSourceSecurityConfig.java
Comment (nishantmonu51) [reply core/src/main/java/org/apache/druid/data/input/impl/InputSourceSecurityConfig.java:45]: done, test is already there but it also had the typo since i copy pasted it, fixed now.

**Review by gianm — COMMENTED** [review]
I think we should have more safeguards to make sure the prefixes are actually getting checked. It would be really easy for someone in the future to implement an InputSource that doesn't check them, or refactor an existing one, and thereby break the security expectations. It would also be easy to add a new parameter to InputSourceSecurityConfig, and forget to update some InputSource to check it.

One possible safeguard: in the docs, explicitly list out which InputSources respect allowPrefixList, add tests for those InputSources to verify it works, and additionally add code to call sites to refuse to run any _other_ InputSource if these params are set. (Otherwise, those InputSources would be a security hole.)

Another: Add a method to InputSource like `validateAllowDenyPrefixList(InputSourceSecurityConfig)` that they must all implement. (This prevents anyone from forgetting it.) Then have callers call that method before using the InputSource to do anything. Add tests for call sites to make sure that method is being called. As an additional safeguard against the _call sites_ forgetting to call the method, have the InputSources throw exceptions in their `reader` methods if the `validateAllowDenyPrefixList` method hasn't previously been called on that same instance.

Thinking about both of these, IMO the second approach is best, because it means that all InputSources will work properly with this parameter. Note: the method must be `validateAllowDenyPrefixList`, not simply `validate`, because if we add a new thing to validate to InputSourceSecurityConfig then we'll need to add a new method to ensure every InputSource is working with it.

File: core/src/main/java/org/apache/druid/data/input/impl/InputSourceSecurityConfig.java
Comment (abhishekagarwal87) [inline core/src/main/java/org/apache/druid/data/input/impl/InputSourceSecurityConfig.java:71]: should return only if there is a match.

File: core/src/test/java/org/apache/druid/data/input/impl/InputSourceSecurityConfigTest.java
Comment (abhishekagarwal87) [inline core/src/test/java/org/apache/druid/data/input/impl/InputSourceSecurityConfigTest.java:47]: I think there should be similar tests but with more than one item on the list.

File: docs/configuration/index.md
Comment (nishantmonu51) [reply docs/configuration/index.md:1363]: yes, they were never released and just committed to master. 
I removed them since having a generic configs that apply would to all inputSources would be easier to configure instead of having source specific configs. .

File: core/src/main/java/org/apache/druid/data/input/impl/InputSourceSecurityConfig.java
Comment (nishantmonu51) [reply core/src/main/java/org/apache/druid/data/input/impl/InputSourceSecurityConfig.java]: done.

File: docs/configuration/index.md
Comment (nishantmonu51) [reply docs/configuration/index.md:1363]: added.

File: core/src/main/java/org/apache/druid/data/input/impl/InputSourceSecurityConfig.java
Comment (nishantmonu51) [reply core/src/main/java/org/apache/druid/data/input/impl/InputSourceSecurityConfig.java]: fixed. thanks for catching it up. Added a test as well.

File: core/src/test/java/org/apache/druid/data/input/impl/InputSourceSecurityConfigTest.java
Comment (nishantmonu51) [reply core/src/test/java/org/apache/druid/data/input/impl/InputSourceSecurityConfigTest.java:48]: added test.

**Comment by nishantmonu51** [comment]
@gianm : Updated the PR based on your feedback and followed the approach of adding a new method to InputSource `validateAllowDenyPrefixList(InputSourceSecurityConfig)`. Please check

File: extensions-core/hdfs-storage/src/main/java/org/apache/druid/inputsource/hdfs/HdfsInputSource.java
Comment (jihoonson) [inline extensions-core/hdfs-storage/src/main/java/org/apache/druid/inputsource/hdfs/HdfsInputSource.java:227]: I think you won't probably need to cache paths since this will unnecessarily populate the cache in the Overlord.

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java
Comment (jihoonson) [inline indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java:158]: Why not `InputSourceSecurityConfig.ALLOW_ALL`? The system-wide securityConfig is ignored in DruidInputSource anyway.

**Review by jihoonson — COMMENTED** [review]
Thanks @nishantmonu51, I left a couple of comments. Also, please fix the CI failures. They look legitimate.

**Comment by jihoonson** [comment]
@nishantmonu51 BTW, since we are about to cut a branch for 0.21.0 release (scheduled on the next Monday), do you think we can get this PR merged before the branch cut? I'm asking because this change will introduce an incompatible change after the release.

File: core/src/main/java/org/apache/druid/data/input/AbstractInputSource.java
Comment (nishantmonu51) [reply core/src/main/java/org/apache/druid/data/input/AbstractInputSource.java:34]: equals tests on base classes were failing without making it transient as they were not including this field in the comparison

File: extensions-core/hdfs-storage/src/main/java/org/apache/druid/inputsource/hdfs/HdfsInputSource.java
Comment (nishantmonu51) [reply extensions-core/hdfs-storage/src/main/java/org/apache/druid/inputsource/hdfs/HdfsInputSource.java:227]: done.

**Comment by nishantmonu51** [comment]
@jihoonson thanks for the review, I hv updated the PR, yes, it will be nice to get this in 0.21.0 release.

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java
Comment (nishantmonu51) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java]: done.

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java
Comment (nishantmonu51) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java:1052]: isReady is also called multiple times, so not sure if that is any better in terms of number of checks, 

Just to double check, if i move it to isReady the task will be accepted by the overlord and it will fail when the task is actually tried to be run. I was thinking that if someone is not allowed, we should block the ingestion at that time when the task is submitted at first place. Is that the behavior you are expecting ?

Comment (jihoonson) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java]: > Just to double check, if i move it to isReady the task will be accepted by the overlord and it will fail when the task is actually tried to be run. I was thinking that if someone is not allowed, we should block the ingestion at that time when the task is submitted at first place. Is that the behavior you are expecting ?

@nishantmonu51 `Task.isReady()` is actually called as soon as it is issued to the Overlord. When it returns true, the Overlord changes its state from `waiting` to `pending`, and tries to assign it to one of middleManagers. So, I think it will be effectively the same as what you want to achieve.

> isReady is also called multiple times, so not sure if that is any better in terms of number of checks,

Good point. I think it won't be not that expensive, but maybe we can check the `validated` state and avoid further checking if it is already validated?

**Review by jihoonson — COMMENTED** [review]
```
[2021-01-12 20:27:36] [build] [2021-01-12 20:27:36] [autobuild] [ERROR] /opt/src/indexing-service/src/test/java/org/apache/druid/indexing/common/task/batch/parallel/PartialCompactionTest.java:[229,12] constructor Builder in class org.apache.druid.indexing.common.task.CompactionTask.Builder cannot be applied to given types;
[2021-01-12 20:27:36] [build] [2021-01-12 20:27:36] [autobuild] [ERROR]   required: java.lang.String,org.apache.druid.indexing.common.SegmentLoaderFactory,org.apache.druid.indexing.common.RetryPolicyFactory
[2021-01-12 20:27:36] [build] [2021-01-12 20:27:36] [autobuild] [ERROR]   found: java.lang.String,org.apache.druid.indexing.common.SegmentLoaderFactory,org.apache.druid.indexing.common.RetryPolicyFactory,org.apache.druid.data.input.impl.InputSourceSecurityConfig
[2021-01-12 20:27:36] [build] [2021-01-12 20:27:36] [autobuild] [ERROR]   reason: actual and formal argument lists differ in length
[2021-01-12 20:27:36] [build] [2021-01-12 20:27:36] [autobuild] [ERROR] /opt/src/indexing-service/src/test/java/org/apache/druid/indexing/common/task/ClientCompactionTaskQuerySerdeTest.java:[197,44] constructor Builder in class org.apache.druid.indexing.common.task.CompactionTask.Builder cannot be applied to given types;
[2021-01-12 20:27:36] [build] [2021-01-12 20:27:36] [autobuild] [ERROR]   required: java.lang.String,org.apache.druid.indexing.common.SegmentLoaderFactory,org.apache.druid.indexing.common.RetryPolicyFactory
[2021-01-12 20:27:36] [build] [2021-01-12 20:27:36] [autobuild] [ERROR]   found: java.lang.String,org.apache.druid.indexing.common.SegmentLoaderFactory,org.apache.druid.indexing.common.RetryPolicyFactory,org.apache.druid.data.input.impl.InputSourceSecurityConfig
[2021-01-12 20:27:36] [build] [2021-01-12 20:27:36] [autobuild] [ERROR]   reason: actual and formal argument lists differ in length
```

@nishantmonu51 thanks for updating the PR. Please check the CI failure. The build failure seems legit.

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java
Comment (nishantmonu51) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java:244]: fixed.

**Comment by nishantmonu51** [comment]
@jihoonson handled review comments, waiting on the CI now.

File: extensions-core/azure-extensions/src/test/java/org/apache/druid/data/input/azure/AzureInputSourceTest.java
Comment (jihoonson) [inline extensions-core/azure-extensions/src/test/java/org/apache/druid/data/input/azure/AzureInputSourceTest.java:220]: Is this the same as the `testDenyAll()` test below?

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamSamplerSpec.java
Comment (jihoonson) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamSamplerSpec.java:109]: nit: this doesn't look harm because neither `RecordSupplierInputSource` nor `FirehoseFactoryToInputSourceAdaptor` do anything in `validateAllowDenyPrefixList()`. However, it looks confusing since `validateAllowDenyPrefixList()` is not called in `RecordSupplier`. I would suggest to delete this line and add it back when we truly support this security config for stream ingestion.

**Review by jihoonson — COMMENTED** [review]
@nishantmonu51 thanks. I left some nit comments. It's up to you to address them or not in this PR.

The CI still fails. At least some of them look legit. I'm not sure about integration test failures, but please take a look at them. Some CI failures are about test coverage. I haven't checked the individual test coverage failures, but I see missing tests for `IndexTask.isReady()` and `ParallelIndexSupervisorTask.isReady()` which I think we should add. You can add some in `IndexTaskTest` and `ParallelIndexSupervisorTaskTest`.

BTW, force-pushing can make the review take longer since reviewers should read the entire change every time. I would suggest to not use it.

**Comment by jihoonson** [comment]
@nishantmonu51 I'm going to cut the branch for 0.21.0 by tonight as we are already behind the schedule (it was originally scheduled on the last Monday). I'm not sure whether we can make it or not because of the test failures. In this case, I would rather not rush but schedule this PR for a future release.

File: extensions-core/azure-extensions/src/test/java/org/apache/druid/data/input/azure/AzureInputSourceTest.java
Comment (nishantmonu51) [reply extensions-core/azure-extensions/src/test/java/org/apache/druid/data/input/azure/AzureInputSourceTest.java]: done.

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamSamplerSpec.java
Comment (nishantmonu51) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamSamplerSpec.java]: done.

**Comment by jihoonson** [comment]
@nishantmonu51 sorry to say, but I created a release branch yesterday. As this PR is unlikely backported (it's a feature and we backport only bug fixes since features can have bugs), I'm going to untag 0.21.0 milestone. This PR will be included in the next milestone. Since this PR didn't make it for 0.21.0, I think we cannot get rid of `HttpInputSourceConfig` immediately but keep it as deprecated for a while. I will not call out `HttpInputSourceConfig` in the release notes to discourage people from using it.

**Comment by jihoonson** [comment]
Hi @nishantmonu51, any updates on this PR? I believe this will be useful from security perspective.

**Comment by nishantmonu51** [comment]
@jihoonson : I will resolve the conflicts and update the PR this week.

**Comment by jihoonson** [comment]
@nishantmonu51 thank you!

**Comment by nishantmonu51** [comment]
merged master and resolved conflicts.

File: docs/configuration/index.md
Comment (jihoonson) [inline docs/configuration/index.md:1528]: I find this section confusing because the `Indexer` here means the node type `Indexer not ingestion. Can you move the new docs to [Ingestion Security Configuration](https://github.com/apache/druid/blob/master/docs/configuration/index.md#ingestion-security-configuration)?

**Review by jihoonson — COMMENTED** [review]
@nishantmonu51 thanks for resolving conflicts. Sadly, there was another PR recently that caused another conflicts. Could you please fix them? Besides conflicts, I have one more comment. The new `InputSourceSecurityConfig` seems to cover all functionalities of `HttpInputSourceConfig` and `HdfsInputSourceConfig`. I think we should deprecate them and encourage people to use the new `InputSourceSecurityConfig` instead. Can you please deprecate those 2 classes and document that people are recommended to use `InputSourceSecurityConfig` instead of them? I think this PR is good to go once my last comments are addressed.

**Comment by jihoonson** [comment]
> [ERROR] org.apache.druid.inputsource.hdfs.HdfsInputSourceTest$ReaderTest.testAllowPath(org.apache.druid.inputsource.hdfs.HdfsInputSourceTest$ReaderTest)
> [ERROR]   Run 1: HdfsInputSourceTest$ReaderTest.testAllowPath:401 » NullPointer inputSourceConf...
> [ERROR]   Run 2: HdfsInputSourceTest$ReaderTest.testAllowPath:401 » NullPointer inputSourceConf...
> [ERROR]   Run 3: HdfsInputSourceTest$ReaderTest.testAllowPath:401 » NullPointer inputSourceConf...
> [ERROR]   Run 4: HdfsInputSourceTest$ReaderTest.testAllowPath:401 » NullPointer inputSourceConf...

> [ERROR] org.apache.druid.metadata.input.InputSourceModuleTest.testHttpInputSourceDefaultConfig(org.apache.druid.metadata.input.InputSourceModuleTest)
> [ERROR]   Run 1: InputSourceModuleTest.testHttpInputSourceDefaultConfig:121 » Configuration Gui...
> [ERROR]   Run 2: InputSourceModuleTest.testHttpInputSourceDefaultConfig:121 » Configuration Gui...
> [ERROR]   Run 3: InputSourceModuleTest.testHttpInputSourceDefaultConfig:121 » Configuration Gui...
> [ERROR]   Run 4: InputSourceModuleTest.testHttpInputSourceDefaultConfig:121 » Configuration Gui...

Oh, I missed the CI failure. Those failed jobs seem legit as seen above.

**Pull request closed without merging**
